### PR TITLE
Add normalisation functions for HF solver

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR adds functions to compute normalisation constant for primitive and contracted Gaussian functions. 

**Description of the Change:**
`primitive_norm` and `contracted_norm` are added to compute normalisation constant for primitive and contracted Gaussian functions, respectively

**Benefits:**
These functions are necessary for computing accurate integrals over basis functions.

**Possible Drawbacks:**
No specific drawbacks.

**Related GitHub Issues:**
No related issues.

### CODE REVIEW

File: pennylane/hf/integrals.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        n (array[float]): normalization coefficient
+
+    **Example**
+    >>> l = (0, 0, 0)
+    >>> alpha = np.array([3.425250914])
+    >>> n = gaussian_norm(l, alpha)
+    >>> print(n)
+    array([1.79444183])
+    """
+    lx, ly, lz = l
```
Comment (agran2018) [inline pennylane/hf/integrals.py:54]: Probably it is easier to change the notation in the docstring for consistency. For example:

> G = x^l y^m z^n e^{-\alpha r^2}  ==>> G = x^l_x y^l_y z^l_z e^{-\alpha r^2}

File: pennylane/hf/integrals.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    n = (
+        (2 * alpha / np.pi) ** 0.75
+        * (4 * alpha) ** (sum(l) / 2)
+        / anp.sqrt(fac2(2 * lx - 1) * fac2(2 * ly - 1) * fac2(2 * lz - 1))
```
Comment (agran2018) [inline pennylane/hf/integrals.py:58]: so many different Numpies :-)

File: pennylane/hf/integrals.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+def contracted_norm(l, alpha, a):
+    r"""Compute the normalization constant for a contracted Gaussian function.
+
+    A contracted Gaussian function is defined as
+
+    .. math::
+
+        \psi = a_1 G_1 + a_2 G_2 + a_3 G_3,
+
+    where :math:`a` denotes the contraction coefficients and :math:`G` is a Gaussian function. The
```
Comment (agran2018) [inline pennylane/hf/integrals.py:72]:
```suggestion
    where :math:`a` denotes the contraction coefficients and :math:`G` is a primitive Gaussian function. The
```

File: pennylane/hf/integrals.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    where :math:`l`, :math:`\alpha` and :math:`a` are the angular momentum quantum numbers, the
+    exponents of the Gaussian functions and the contraction coefficients, respectively.
+
+    Args:
+        l (tuple[int]): angular momentum quantum numbers of the basis function
+        alpha (array[float]): exponents of the primitive Gaussian functions
+        a (array[float]): coefficients of the contracted Gaussian functions
+
+    Returns:
+        n (array[float]): normalization coefficient
```
Comment (agran2018) [inline pennylane/hf/integrals.py:89]: same comment here, any reason for `n` to be an array?

File: pennylane/hf/integrals.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    >>> n = contracted_norm(l, alpha, a)
+    >>> print(n)
+    0.39969026908800853
+    """
+    lx, ly, lz = l
+    c = np.pi ** 1.5 / 2 ** sum(l) * fac2(2 * lx - 1) * fac2(2 * ly - 1) * fac2(2 * lz - 1)
+    s = (
+        (a.reshape(len(a), 1) * a) / ((alpha.reshape(len(alpha), 1) + alpha) ** (sum(l) + 1.5))
+    ).sum()
+    n = 1 / anp.sqrt(c * s)
```
Comment (agran2018) [inline pennylane/hf/integrals.py:104]: Indeed, it is not an array.

**Review by agran2018 — CHANGES_REQUESTED** [review]
Thank you @soranjh! Left some suggestions to improve consistency of the code with the docstrings.

File: pennylane/hf/integrals.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    normalization constant for this function is computed as
+
+    .. math::
+
+        N(l, \alpha) = (\frac{2\alpha}{\pi})^{3/4} \frac{(4 \alpha)^{(l_x + l_y + l_z)/2}}
+        {(2\l_x-1)!! (2\l_y-1)!! (2\l_z-1)!!)^{1/2}}.
+
+    Args:
+        l (tuple[int]): angular momentum quantum numbers of the basis function
+        alpha (array[float]): exponent of the primitive Gaussian function
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/hf/integrals.py:42]: When `\alpha` is a differentiable parameter it has the format of a pennylane numpy array and each component of the array has the same format. 
```
from pennylane import numpy as pnp
alpha = pnp.array([3.42525091, 0.62391373, 0.1688554])
alpha[0]  # tensor(3.42525091, requires_grad=True)
```
Same applies to the rest of occasions pointed to below.

**Review by agran2018 — APPROVED** [review]
Thank you @soranjh. Looks good on my side.

File: pennylane/hf/integrals.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def primitive_norm(l, alpha):
+    r"""Compute the normalization constant for a primitive Gaussian function.
+
+    A Gaussian function centred at the position :math:`r = (x, y, z)` is defined as
+
+    .. math::
+
+        G = x^l_x y^l_y z^l_z e^{-\alpha r^2},
```
Comment (rmoyard) [inline pennylane/hf/integrals.py:29]:
```suggestion
        G = x^{l_x} y^{l_y} z^{l_z} e^{-\alpha r^2},
```
I am not sure it renders correctly without the {}.

File: pennylane/hf/integrals.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    Returns:
+        n (array[float]): normalization coefficient
+
+    **Example**
+    >>> l = (0, 0, 0)
+    >>> alpha = np.array([3.425250914])
+    >>> n = gaussian_norm(l, alpha)
+    >>> print(n)
+    array([1.79444183])
```
Comment (rmoyard) [inline pennylane/hf/integrals.py:51]:
```suggestion
    **Example**
    
    >>> l = (0, 0, 0)
    >>> alpha = np.array([3.425250914])
    >>> n = gaussian_norm(l, alpha)
    >>> print(n)
    
    array([1.79444183])
```

File: pennylane/hf/integrals.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        G = x^l y^m z^n e^{-\alpha r^2},
+
+    where :math:`l = (l, m, n)` defines the angular momentum quantum numbers, :math:`\alpha`
+    is the exponent and :math:`r = (x, y, z)` determines the center of the function. The
+    normalization constant for this function is computed as
+
+    .. math::
+
+        N(l, \alpha) = (\frac{2\alpha}{\pi})^{3/4} \frac{(4 \alpha)^{(l_x + l_y + l_z)/2}}
+        {(2\l_x-1)!! (2\l_y-1)!! (2\l_z-1)!!)^{1/2}}.
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/hf/integrals.py:37]: Is that correct to use \l ?

File: pennylane/hf/integrals.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    .. math::
+
+        \psi = a_1 G_1 + a_2 G_2 + a_3 G_3,
+
+    where :math:`a` denotes the contraction coefficients and :math:`G` is a primitive Gaussian function. The
+    normalization constant for this function is computed as
+
+    .. math::
+
+        N(l, \alpha, a) = [\frac{\pi^{3/2}(2\l_x-1)!! (2\l_y-1)!! (2\l_z-1)!!}{2^{l_x + l_y + l_z}}
```
Comment (rmoyard) [inline pennylane/hf/integrals.py:76]: Same comment as before with \l.

File: pennylane/hf/integrals.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    Returns:
+        n (array[float]): normalization coefficient
+
+    **Example**
+    >>> l = (0, 0, 0)
+    >>> alpha = np.array([3.425250914, 0.6239137298, 0.168855404])
+    >>> a = np.array([1.79444183, 0.50032649, 0.18773546])
+    >>> n = contracted_norm(l, alpha, a)
+    >>> print(n)
```
Comment (rmoyard) [inline pennylane/hf/integrals.py:95]:
```suggestion
    **Example**
    
    >>> l = (0, 0, 0)
    >>> alpha = np.array([3.425250914, 0.6239137298, 0.168855404])
    >>> a = np.array([1.79444183, 0.50032649, 0.18773546])
    >>> n = contracted_norm(l, alpha, a)
    >>> print(n)
    
```

**Review by rmoyard — COMMENTED** [review]
Great PR, I've left some comments about the docstrings
